### PR TITLE
Add parser generators which handles RssTag and RssAttribute.

### PR DESCRIPTION
**Pull request by ivanisidrowu**
## Intro
This code generator could be a little bit complicated.
I use KotlinPoet to generate code. Please read the document first.
https://square.github.io/kotlinpoet

### Data
We assume that a user defines a data model like this one.

```kotlin
@RssTag(name = "channel")
data class RssData(
    val title: String?,
    val textInput: MyTextInput?,
    @RssTag(name = "item")
    val list: List<RssItem>,
    val cloud: Cloud?,
    @RssTag(name = "category")
    val categories: List<Category>,
): Serializable

@RssTag(name = "textInput")
data class MyTextInput(
    val title: String?,
    val name: String?
): Serializable

@RssTag(name = "item")
data class RssItem(
    val title: String?,
    val author: String?,
): Serializable

@RssTag(name = "cloud")
data class Cloud(
    @RssAttribute
    val domain: String?,
    @RssAttribute(name = "port")
    val portNum: Int?,
    @RssAttribute
    val path: String?
): Serializable

@RssTag(name = "category")
data class Category(
    @RssAttribute
    val domain: String?
)
```

### Generated Extensions

First, we will generate a set of extensions for future usage.

```kotlin
object ParserExtensions {
  fun getXmlParser(xml: String): XmlPullParser {

    ByteArrayInputStream(xml.toByteArray()).use { inputStream ->
        return Xml.newPullParser().apply {
            setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
            setInput(inputStream, null)
            nextTag()
        }
    }
  }

  fun XmlPullParser.readString(tagName: String): String? {
    require(XmlPullParser.START_TAG, null, tagName)
    var content: String? = null
    if (next() == XmlPullParser.TEXT) {
    content = text
    nextTag()
        if (eventType != XmlPullParser.END_TAG) {
          skip()
          nextTag()
          content = null
        }
    }
    require(XmlPullParser.END_TAG, null, tagName)
    return content
  }

  fun XmlPullParser.skip() {
    if (eventType != XmlPullParser.START_TAG) {
      throw IllegalStateException()
    }
    var depth = 1
    while (depth != 0) {
        when (next()) {
            XmlPullParser.END_TAG -> depth--
            XmlPullParser.START_TAG -> depth++
        }
    }
  }

  fun String.toBooleanOrNull(): Boolean? = when (toLowerCase()) {
      "true" -> true
      "false" -> false
      "yes" -> true
      "no" -> false
      else -> null
  }}
```

### Generated Parser

In this case, we will generate 5 parser classes, such as `RssDataParser`, `MyTextInputParser`, `RssItemParser`, `CloudParser`, and `CategoryParser`.

In general, the generated parser will look like this one.

```kotlin
object RssDataParser {
    fun parse(xml: String): RssData {
        val parser = getXmlParser(xml)

        var result: RssData? = null
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) continue

            if (parser.name == "channel") {
                result = parser.getChannel()
                break
            } else {
                parser.skip()
            }
        }
        return result ?: throw XmlPullParserException("No valid channel tag in the RSS feed.")
    }

    fun XmlPullParser.getChannel(): RssData {
        require(XmlPullParser.START_TAG, null, "channel")
        var title: String? = null
        var textInput: MyTextInput? = null
        var list: ArrayList<RssItem> = arrayListOf()
        var cloud: Cloud? = null
        var categories: ArrayList<Category> = arrayListOf()
        while (next() != XmlPullParser.END_TAG) {
            if (eventType != XmlPullParser.START_TAG) continue

            when (this.name) {
                "title" -> title = readString("title")
                "textInput" -> textInput = getTextInput()
                "item" -> list.add(getItem())
                "cloud" -> cloud = getCloud()
                "category" -> categories.add(getCategory())
                else -> skip()
            }
        }
        require(XmlPullParser.END_TAG, null, "channel")
        return RssData(
            title = title,
            textInput = textInput,
            list = list,
            cloud = cloud,
            categories = categories,
        )
    }
}
```

## TODO
- Add Log (in compile time) to the generated parser.
- Add an API generator.
- Handle `@RssRawData` and tag orders.
- Add tests.

### CODE REVIEW

File: processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                outputClass, CHANNEL, xmlParserExceptionClass, skipMemberName, getParserMemberName
+            )
+            .returns(outputClass)
+            .build()
+    }
+
+    private fun getClassFunSpec(rootElement: Element, outputClassName: String): FunSpec {
+        val outputClass = ClassName(rootElement.getPackage(), outputClassName)
+        val annotation = rootElement.getAnnotation(RssTag::class.java)
+        val tagName = annotation?.name?.takeIfNotEmpty() ?: rootElement.simpleName.toString()
```
Comment (dev-weiqi) [inline processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt:91]: `val tagName = annotation?.name?.ifEmpty { rootElement.simpleName.toString() }`

Comment (ivanisidrowu) [reply processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt:91]: 這樣 `tagName` 有可能會從 `String` 變成 `String?` 喔！

File: processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        } else {
+            addStatement(statement, readStringMemberName)
+        }
+    }
+
+    private fun String.appendTypeConversion(typeString: String): String {
+        return when (typeString) {
+            String::class.java.simpleName -> this
+            Integer::class.java.simpleName -> "$this.toIntOrNull()"
+            Boolean::class.java.simpleName -> "$this.%2M()"
```
Comment (HsienDaDa) [inline processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt:282]: %2M()  這是啥？

Comment (ivanisidrowu) [reply processor/src/main/java/tw/ktrssreader/processor/ParserGenerator.kt:282]: KotlinPoet 中的第二個 member name
https://square.github.io/kotlinpoet/#m-for-members

**Review by dev-weiqi — COMMENTED** [review]
done!

**Review by HsienDaDa — APPROVED** [review]

File: processor/src/main/java/tw/ktrssreader/processor/util/Extensions.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+fun String.isListType(): Boolean {
+    return contains(List::class.java.canonicalName)
+}
+
+fun String.isPrimitive(): Boolean {
+    primitiveJavaPaths.forEach {
+        if (this.contains(it)) return true
+    }
+    return false
```
[the remark this answers is not in the document]
Comment (ivanisidrowu) [reply processor/src/main/java/tw/ktrssreader/processor/util/Extensions.kt]: done

**Review by dev-weiqi — APPROVED** [review]

**Pull request merged**